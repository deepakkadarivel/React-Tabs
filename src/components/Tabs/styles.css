.Tabs {
  list-style: none;
  display: flex;
  border-bottom: 2px solid #cccccc;
  margin: 24px 8px;
  padding: 0;
}

.Tabs-Item {
  padding: 8px 16px;
  margin: 0 4px;
  letter-spacing: 0.6px;
  font-size: 14px;
}

.Tabs-Item > a {
  color: #888888;
  text-decoration: none;
}

.Tabs-Item--current {
  font-weight: 500;
  background: #ffffff;
  border: 2px solid #cccccc;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border-bottom: 0;
  position: relative;
  top: 2px;
}

.Tabs-Item--current > a {
  color: #000000;
  text-decoration: none;
}
