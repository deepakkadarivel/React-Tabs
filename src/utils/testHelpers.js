
import { render } from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

const renderWithRouter = (ui, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return render(ui, { wrapper: BrowserRouter })
}

const leftClick = { button: 0 }
const  click =(ui)=> userEvent.click(ui, leftClick)

export { renderWithRouter, click };