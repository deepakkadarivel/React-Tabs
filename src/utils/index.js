import { useQuery } from "./router";
import { renderWithRouter, click } from "./testHelpers";
export { useQuery , renderWithRouter, click};
