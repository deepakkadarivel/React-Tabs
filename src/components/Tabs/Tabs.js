import React, { useCallback, useEffect, useState } from "react";
import PropTypes from 'prop-types';
import cx from "classnames";
import { useRouteMatch, Link } from "react-router-dom";
import { useQuery } from "../../utils";
import "./styles.css";

const Tabs = ({ content }) => {
  let match = useRouteMatch();
  let query = useQuery();
  const tab = query.get("tab");

  const [activeTab, setActiveTab] = useState(null);

  const handleTabChange = useCallback(() => {
    const selectedTab = content.find((t) => t.key === tab) || content[0];
    setActiveTab(selectedTab);
  }, [tab, content]);

  useEffect(() => {
    handleTabChange();
  }, [handleTabChange]);

  return (
    <div>
      <ul className="Tabs">
        {content?.map((tab) => {
          const isActive = activeTab?.key === tab.key;
          return (
            <li
              data-testid={`t-${tab.key}`}
              className={cx("Tabs-Item", { "Tabs-Item--current": isActive })}
              key={tab.key}
            >
              <Link to={`${match.path}?tab=${tab.key}`}>{tab.name}</Link>
            </li>
          );
        })}
      </ul>

      <div data-testid="t-tabContent" className="Tab-Content">{activeTab?.content}</div>
    </div>
  );
};

Tabs.propTypes = {
  content: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    name: PropTypes.string,
    content: PropTypes.any
  })).isRequired
}

Tabs.defaultProps = {
  content: []
}

export { Tabs };
