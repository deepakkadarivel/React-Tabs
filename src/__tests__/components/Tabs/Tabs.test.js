import React from "react";
import '@testing-library/jest-dom';
import { Tabs } from "../../../components/Tabs";
import { renderWithRouter } from '../../../utils';

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useRouteMatch: () => ({ url: "/", path: "/" }),
  useLocation: () => ({ search: 'tab1'}),
}));

const tabs = [
  { name: "Tab 1", key: "tab1", content: <span>ABC</span> },
  { name: "Tab 2", key: "tab2", content: <span>DEF</span> },
  { name: "Tab 3", key: "tab3", content: <span>GHI</span> }
];

it("renders <Tabs />", () => {
  const { container, getByTestId } = renderWithRouter(<Tabs content={tabs} />);
  
  // Snapshot
  expect(container).toMatchSnapshot();

  // Tab 1
  const tab1 = getByTestId('t-tab1');
  expect(tab1).toHaveClass('Tabs-Item Tabs-Item--current');
  expect(tab1.textContent).toBe('Tab 1');
  
  // Tab 2
  const tab2 = getByTestId('t-tab2');
  expect(tab2).toHaveClass('Tabs-Item');
  expect(tab2.classList).not.toContain('Tabs-Item--current');
  expect(tab2.textContent).toBe('Tab 2');
  
  // Tab 3
  const tab3 = getByTestId('t-tab3');
  expect(tab3).toHaveClass('Tabs-Item');
  expect(tab3.classList).not.toContain('Tabs-Item--current');
  expect(tab3.textContent).toBe('Tab 3');

  // Tab content
  const content = getByTestId('t-tabContent');
  expect(content.textContent).toBe('ABC');
});

it('renders <Tabs /> without crashing.', () => {
  const div = document.createElement('div');
  renderWithRouter(<Tabs />, div);
})