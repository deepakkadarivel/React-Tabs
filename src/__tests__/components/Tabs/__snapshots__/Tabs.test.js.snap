// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Tabs /> 1`] = `
<div>
  <div>
    <ul
      class="Tabs"
    >
      <li
        class="Tabs-Item Tabs-Item--current"
        data-testid="t-tab1"
      >
        <a
          href="/?tab=tab1"
        >
          Tab 1
        </a>
      </li>
      <li
        class="Tabs-Item"
        data-testid="t-tab2"
      >
        <a
          href="/?tab=tab2"
        >
          Tab 2
        </a>
      </li>
      <li
        class="Tabs-Item"
        data-testid="t-tab3"
      >
        <a
          href="/?tab=tab3"
        >
          Tab 3
        </a>
      </li>
    </ul>
    <div
      class="Tab-Content"
      data-testid="t-tabContent"
    >
      <span>
        ABC
      </span>
    </div>
  </div>
</div>
`;
